tests/test_fidelity.py

```diff
@@ -5,85 +5,125 @@
+
+    @pytest.mark.parametrize("hbar", [0.5, 1.0, 2.0, 1.6])
+    @pytest.mark.parametrize("r1", np.random.rand(3))
+    @pytest.mark.parametrize("r2", np.random.rand(3))
+    def test_fidelity_squeezed_vacuum(self, r1, r2, hbar):
+        """Tests fidelity between two squeezed states"""
+        cov1 = np.diag([np.exp(2 * r1), np.exp(-2 * r1)]) * hbar / 2
+        cov2 = np.diag([np.exp(2 * r2), np.exp(-2 * r2)]) * hbar / 2
+        mu = np.zeros([2])
+        assert np.allclose(1 / np.cosh(r1 - r2), gp.fidelity(mu, cov1, mu, cov2, hbar=hbar))
+
+    @pytest.mark.parametrize("n1", [0.5, 1.0, 2.0, 1.6])
+    @pytest.mark.parametrize("n2", [0.5, 1.0, 2.0, 1.6])
+    @pytest.mark.parametrize("hbar", [0.5, 1.0, 2.0, 1.6])
+    def test_fidelity_thermal(self, n1, n2, hbar):
+        """Test fidelity between two thermal states"""
+        expected = 1 / (1 + n1 + n2 + 2 * n1 * n2 - 2 * np.sqrt(n1 * n2 * (n1 + 1) * (n2 + 1)))
+        cov1 = hbar * (n1 + 0.5) * np.identity(2)
+        cov2 = hbar * (n2 + 0.5) * np.identity(2)
+        mu1 = np.zeros([2])
+        mu2 = np.zeros([2])
+        assert np.allclose(expected, gp.fidelity(mu1, cov1, mu2, cov2, hbar=hbar))
+
+    @pytest.mark.parametrize("hbar", [0.5, 1.0, 2.0, 1.6])
+    @pytest.mark.parametrize("r", [-2.0, 0.0, 2.0])
+    @pytest.mark.parametrize("alpha", np.random.rand(10) + 1j * np.random.rand(10))
+    def test_fidelity_vac_to_displaced_squeezed(self, r, alpha, hbar):
+        """Calculates the fidelity between a coherent squeezed state and vacuum"""
+        cov1 = np.diag([np.exp(2 * r), np.exp(-2 * r)]) * hbar / 2
+        means1 = real_to_complex_displacements(np.array([alpha, np.conj(alpha)]), hbar=hbar)
+        means2 = np.zeros([2])
+        cov2 = np.identity(2) * hbar / 2
+        expected = (
+            np.exp(-np.abs(alpha) ** 2)
+            * np.abs(np.exp(np.tanh(r) * np.conj(alpha) ** 2))
+            / np.cosh(r)
+        )
+        assert np.allclose(expected, gp.fidelity(means1, cov1, means2, cov2, hbar=hbar))
+
+
+class TestMixedStates:
+
+    state1 = 1 / 2 * np.eye(2)
+
+    state2 = 1 / 3 * np.ones((2, 2))
+    state2[1, 1] = 2 / 3
+
+    def test_fidelity_with_self(self):
+        """Test that the fidelity of two identical quantum states is 1"""
+        assert np.allclose(fp.fidelity(self.state1, self.state1, False, False), 1, atol=1e-4)
+
+    def test_fidelity_is_symmetric(self):
+        """Test that the fidelity is symmetric and between 0 and 1"""
+        f12 = fp.fidelity(self.state1, self.state2, False, False)
+        f21 = fp.fidelity(self.state2, self.state1, False, False)
+        assert np.allclose(f12, f21)
+        assert 0 <= np.real_if_close(f12) < 1.0
+
+    def test_fidelity_formula(self):
+        """Test fidelity of mixed states agrees with definition.
+
+        Richard Jozsa (1994) Fidelity for Mixed Quantum States,
+        Journal of Modern Optics, 41:12, 2315-2323,
+        DOI: 10.1080/09500349414552171
```

Comment:
Perhaps move this to the implementation instead, since this is the definition used there.

Reply:
Not sure if this test adds much, though. Would be better to have a test with hard-coded values for the fidelity (that we know are correct).

Reply:
Good point! I'll do that